### PR TITLE
UniqueIndex range_de

### PR DESCRIPTION
**Pull request by uint**
Deals with #461

### CODE REVIEW

File: packages/storage-plus/src/indexes.rs
Comment (maurolacy) [inline packages/storage-plus/src/indexes.rs:397]: This code block is correct, but the test below is failing because `K` here is the **unique index** key type (`U32Key` in the test), not the primary key type (`&[u8]` in the test).

I think you will need to modify `UniqueIndex` to pass the primary key type (a `PK` type parameter) to it, and use that when instantiating `deserialize_unique_kv`.

File: packages/storage-plus/src/indexes.rs
Comment (maurolacy) [inline packages/storage-plus/src/indexes.rs:470]:
```suggestion
impl<'a, PK, K, T> UniqueIndex<'a, PK, K, T>
```

File: packages/storage-plus/src/indexes.rs
Comment (maurolacy) [inline packages/storage-plus/src/indexes.rs:473]:
```suggestion
    K: PrimaryKey<'a>,
    PK: KeyDeserialize,
```

and so on.

File: packages/storage-plus/src/indexes.rs
Comment (maurolacy) [inline packages/storage-plus/src/indexes.rs:527]:
```suggestion
    fn no_prefix_de(&self) -> Prefix<PK, T> {
        Prefix::with_deserialization_function(self.idx_namespace, &[], &[], |_, _, kv| {
            deserialize_unique_kv_de::<_, PK>(kv)
```

and so on.

File: packages/storage-plus/src/indexed_map.rs
Comment (maurolacy) [reply packages/storage-plus/src/indexed_map.rs:841]: `Prefix` must work with the key of the index in this case (that's what you are prefixing). In this impl, what must be changed is the return value of `Prefix`. Instead of being `Suffix`, it has to be `PK` (for `UniqueIndex`).

Not sure about this, by the way. Will have to take a detailed look to answer further, in a more detailed or precise way.

What we can do is to leave this on hold, documenting the issues, and take a second attempt at it in the future.

**Comment by ethanfrey** [comment]
I agree this is mergeable.
Some follow up is clear.

Let's find time to discuss architecture for the larger follow-up issues before tackling them (but feel free to make issues in the meantime)

**Comment by ethanfrey** [comment]
Oh, man, there are like 5 PRs built on top of this, right?

**Comment by maurolacy** [comment]
Yes, I'll rebase everything once the int key changes are merged to main.

**Comment by maurolacy** [comment]
> Can you merge main in sometime (no more rebase possible)

Just rebased it without issues.

**Comment by ethanfrey** [comment]
Cool. Want to merge this first, then the other branches on top?

**Comment by maurolacy** [comment]
Sure.

**Comment by maurolacy** [comment]
OK, all the follow-up issues / comments are addressed, are changes are merged / integrated here.

This is also rebased from master.

Let's merge this.

File: packages/storage-plus/src/indexes/mod.rs
Comment (uint) [inline packages/storage-plus/src/indexes/mod.rs:16]: These are test helpers, right? This isn't really relevant to the PR, but shouldn't we compile these only in `test` builds?

File: packages/storage-plus/src/indexed_snapshot.rs
Comment (ueco-jb) [inline packages/storage-plus/src/indexed_snapshot.rs:239]: I'd be inclined to rephrase this part, personally I'm not sure I follow:
> assumes you set the prefix to one element

meaning assumes that each element has prefix set?
and 
> call range over the last one

meaning call range over last one (element)? So since last element?

**Comment by uint** [comment]
LGTM overall. Does what we need it to!

File: packages/storage-plus/src/indexes/mod.rs
Comment (maurolacy) [reply packages/storage-plus/src/indexes/mod.rs:16]: Those are helpers that _could_ be used from contract code. They are in fact exported in `lib.rs`. They are not very general, but still.

**Review by ueco-jb — APPROVED** [review]

File: packages/storage-plus/src/indexed_snapshot.rs
Comment (maurolacy) [reply packages/storage-plus/src/indexed_snapshot.rs:239]: The difference is somewhat subtle: `range` over a `prefix` only iterates over all the elements with **that** prefix. Whereas `prefix_range` allows you to set the range directly over the prefix. Meaning prefix can now vary (inclusively or exclusively, depending on `PrefixBound`) between the `min` and `max` `prefix_range` parameters, if specified.

Comment (maurolacy) [reply packages/storage-plus/src/indexed_snapshot.rs:239]: I'll try to clarify this in the comment.

**Pull request merged**
